File: apps/web/src/stores/auth-store.ts
import { create } from 'zustand'
import { persist } from 'zustand/middleware'
import { authApi, User } from '@/lib/auth-api'

// 声明 FlutterChannel 类型
declare global {
  interface Window {
    FlutterChannel?: { postMessage: (msg: string) => void }
  }
}

const TOKEN_KEY = 'auth_token'

function consumeQrLoginToken() {
  if (typeof window === 'undefined') return null

  const currentUrl = new URL(window.location.href)
  if (currentUrl.searchParams.get('qrLogin') !== '1') return null

  const qrToken = currentUrl.searchParams.get('token')
  currentUrl.searchParams.delete('qrLogin')
  currentUrl.searchParams.delete('token')
  currentUrl.searchParams.delete('username')
  window.history.replaceState({}, document.title, `${currentUrl.pathname}${currentUrl.search}${currentUrl.hash}`)

  if (qrToken) {
    localStorage.setItem(TOKEN_KEY, qrToken)
  }
  return qrToken
}

export type AuthState = 'checking' | 'unauthenticated' | 'authenticated'

interface AuthStore {
  state: AuthState
  user: User | null
  token: string | null
  isFirstUse: boolean
  login: (username: string, password: string) => Promise<{ success: boolean; error?: string }>
  register: (username: string, password: string, avatar?: string) => Promise<{ success: boolean; error?: string }>
  logout: () => void
  checkAuth: () => Promise<void>
  setUser: (user: User) => void
}

export const useAuthStore = create<AuthStore>()(
  persist(
    (set, _get) => ({
      state: 'checking',
      user: null,
      token: null,
      isFirstUse: false,

      checkAuth: async () => {
        // 移动端 WebView 需要等待 token 注入
        // 检测是否在移动端 WebView 中（通过 UserAgent 或 FlutterChannel）
        const isMobileWebView = typeof window !== 'undefined' &&
          (window.navigator.userAgent.includes('Flutter') ||
           'FlutterChannel' in window);

        if (isMobileWebView) {
          // 等待 Flutter 注入 token
          await new Promise(resolve => setTimeout(resolve, 200));
        }

        const storedToken = consumeQrLoginToken() || localStorage.getItem(TOKEN_KEY)

        if (storedToken) {
          // Validate token with server
          const response = await authApi.me(storedToken)
          if (response.success && response.data) {
            set({
              user: response.data,
              token: storedToken,
              state: 'authenticated',
            })
            localStorage.setItem('auth_user', JSON.stringify(response.data))
          } else {
            // Token invalid, clear storage
            localStorage.removeItem(TOKEN_KEY)
            localStorage.removeItem('auth_user')
            set({
              token: null,
              user: null,
              state: 'unauthenticated',
            })
          }
        } else {
          // No token, check if first use
          const firstUseResponse = await authApi.checkFirstUse()
          if (firstUseResponse.success && firstUseResponse.data) {
            set({ isFirstUse: firstUseResponse.data.isFirstUse })
          }
          set({ state: 'unauthenticated' })
        }
      },

      login: async (username: string, password: string) => {
        const response = await authApi.login({ username, password })
        if (response.success && response.data) {
          localStorage.setItem(TOKEN_KEY, response.data.token)
          localStorage.setItem('auth_user', JSON.stringify(response.data.user))
          set({
            token: response.data.token,
            user: response.data.user,
            state: 'authenticated',
          })
          return { success: true }
        }
        return { success: false, error: response.error || 'Login failed' }
      },

      register: async (
        username: string,
        password: string,
        avatar?: string,
      ) => {
        const response = await authApi.register({
          username,
          password,
          avatar,
        })
        if (response.success && response.data) {
          localStorage.setItem(TOKEN_KEY, response.data.token)
          localStorage.setItem('auth_user', JSON.stringify(response.data.user))
          set({
            token: response.data.token,
            user: response.data.user,
            state: 'authenticated',
          })
          return { success: true }
        }
        return { success: false, error: response.error || 'Registration failed' }
      },

      logout: () => {
        localStorage.removeItem(TOKEN_KEY)
        localStorage.removeItem('auth_user')
        set({
          token: null,
          user: null,
          state: 'unauthenticated',
        })
        // 通知移动端退出登录
        if (typeof window !== 'undefined') {
          // Flutter webview_flutter 的 JavaScriptChannel 直接挂在全局
          try {
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
            ;(window as any).FlutterChannel?.postMessage(JSON.stringify({ type: 'logout' }))
          } catch {
            // FlutterChannel 可能未定义
          }
        }
      },

      setUser: (user: User) => {
        set({ user })
      },
    }),
    {
      name: 'auth-storage',
      partialize: (state) => ({
        // Only persist user info, not sensitive token in zustand persist
        // Token is handled separately via localStorage for auth header
        user: state.user,
      }),
    }
  )
)

File: apps/mobile/lib/stores/auth_store.dart
import 'package:flutter/foundation.dart';
import '../models/models.dart';
import '../services/storage_service.dart';
import '../services/api_client.dart';

/// 认证状态
enum AuthState { checking, unauthenticated, authenticated }

/// 认证状态管理
class AuthStore extends ChangeNotifier {
  User? _user;
  AuthState _authState = AuthState.checking;
  bool _isLoading = false;
  bool _isFirstUse = true;

  User? get user => _user;
  AuthState get authState => _authState;
  bool get isLoading => _isLoading;
  bool get isFirstUse => _isFirstUse;

  /// 检查认证状态
  Future<void> checkAuth() async {
    print('[AuthStore] checkAuth 开始');
    _authState = AuthState.checking;
    _isLoading = true;
    notifyListeners();

    try {
      final token = await StorageService.getToken();
      print('[AuthStore] token: ${token != null ? '存在' : '不存在'}');

      if (token == null) {
        _authState = AuthState.unauthenticated;
        _isLoading = false;
        notifyListeners();
        return;
      }

      final response = await AuthApi.me();
      if (response.success && response.data != null) {
        _user = response.data;
        _authState = AuthState.authenticated;
        _isLoading = false;
        notifyListeners();
      } else {
        await StorageService.deleteToken();
        _authState = AuthState.unauthenticated;
        _isLoading = false;
        notifyListeners();
      }
    } catch (e) {
      print('[AuthStore] checkAuth 异常: $e');
      _authState = AuthState.unauthenticated;
      _isLoading = false;
      notifyListeners();
    }
  }

  /// 检查是否首次使用
  Future<void> checkFirstUse() async {
    try {
      final response = await AuthApi.checkFirstUse();
      if (response.success && response.data != null) {
        _isFirstUse = response.data!['isFirstUse'] ?? false;
        notifyListeners();
      }
    } catch (e) {
      _isFirstUse = false;
    }
  }

  /// 登录
  Future<Map<String, dynamic>> login(String username, String password) async {
    _isLoading = true;
    notifyListeners();

    try {
      final response = await AuthApi.login(
        username: username,
        password: password,
      );

      if (!response.success || response.data == null) {
        _isLoading = false;
        notifyListeners();
        return {'success': false, 'error': response.error ?? '登录失败'};
      }

      final token = response.data!['token'] as String;
      final userData = response.data!['user'] as Map<String, dynamic>;

      await StorageService.saveToken(token);

      _user = User.fromJson(userData);
      _authState = AuthState.authenticated;
      _isLoading = false;
      _isFirstUse = false;
      notifyListeners();

      return {'success': true};
    } catch (e) {
      _isLoading = false;
      notifyListeners();
      return {'success': false, 'error': '网络错误，请检查连接'};
    }
  }

  /// 二维码登录
  Future<Map<String, dynamic>> qrLogin(
    String url,
    String token,
    String username,
  ) async {
    _isLoading = true;
    notifyListeners();

    try {
      await StorageService.setServerUrl(url);
      await StorageService.saveToken(token);

      _user = User(id: '', username: username, createdAt: '');
      _authState = AuthState.authenticated;
      _isLoading = false;
      _isFirstUse = false;
      notifyListeners();

      return {'success': true};
    } catch (e) {
      _isLoading = false;
      notifyListeners();
      return {'success': false, 'error': '保存登录信息失败'};
    }
  }

  /// 注册
  Future<Map<String, dynamic>> register(
    String username,
    String password,
    String? avatar,
  ) async {
    _isLoading = true;
    notifyListeners();

    try {
      final response = await AuthApi.register(
        username: username,
        password: password,
        avatar: avatar,
      );

      if (!response.success || response.data == null) {
        _isLoading = false;
        notifyListeners();
        return {'success': false, 'error': response.error ?? '注册失败'};
      }

      final token = response.data!['token'] as String;
      final userData = response.data!['user'] as Map<String, dynamic>;

      await StorageService.saveToken(token);

      _user = User.fromJson(userData);
      _authState = AuthState.authenticated;
      _isLoading = false;
      _isFirstUse = false;
      notifyListeners();

      return {'success': true};
    } catch (e) {
      _isLoading = false;
      notifyListeners();
      return {'success': false, 'error': '网络错误'};
    }
  }

  /// 退出登录
  Future<void> logout() async {
    await StorageService.deleteToken();
    _user = null;
    _authState = AuthState.unauthenticated;
    notifyListeners();
  }
}